### PR TITLE
Space drifter ghostrole

### PR DESCRIPTION
**Pull request by c4llv07e**
## About the PR

This PR adds a new midround ghost role: space drifter. He appears in a random location at the station and can see and hear(under question) ghosts.

## Why / Balance

I'm not sure if it's a good idea to give players the opportunity to chat with ghosts, because dead chat is more OOC rather than IC, but as some rounds have shown, it provides a great opportunity for RP without breaking the rules.

## Technical details

GhostObserver and DeadHearing components have been added, which allow entities to see and hear ghosts respectively.
Unlike GhostHearingComponent, DeadHearingComponent only hears ghosts in the voice range.

## Media

https://github.com/space-wizards/space-station-14/assets/38111072/827c6d55-bba7-4290-bee5-e037df0f49ca

The localization files in the video are broken, here is the current version:

image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: c4llv07e
- add: Added new midround ghost role: space drifter!

### CODE REVIEW

**Comment by LankLTE** [comment]
Just telling people "you can't use metainfo from ghosts" is probably a poor thing to try and implement since admins now have to enforce what is / isn't metainfo and that's also not super clear for players either.

**Comment by EmoGarbage404** [comment]
I'm willing to experiment with seeing ghosts, but they definitely should not be able to hear deadchat.

**Comment by chromiumboy** [comment]
Hobo is somewhat derogatory, another name would be preferable, e.g., vagrant or drifter. Or you could lean into the ghost aspects and call them a medium, seer, psychic, etc

**Comment by IamVelcroboy** [comment]
> Hobo is somewhat derogatory, another name would be preferable, e.g., vagrant or drifter. Or you could lean into the ghost aspects and call them a medium, seer, psychic, etc

Space Drifter sounds kinda cool

**Comment by IamVelcroboy** [comment]
Or, in the early 1900's they called them "Ramblers"

**Comment by Hmeister-fake** [comment]
Maybe they should hear 1 word from whatever the ghosts say with a 1/3 chance. Like in Delta-v when you sleep you hear psionics.

**Comment by chromiumboy** [comment]
I wouldn't have ghost chat in any capacity being hearable. Too much OOC and discussion of the current game state. And if you just make it less, or entirely, intelligible then it doesn't really add much to the role. You could have a set of single words in a list that ghosts could send to them maybe, like 'yes', 'no', 'danger' etc. But you'd want it to be either vote based or on a cooldown to prevent spamming

**Comment by Aisu9** [comment]
please for the love of god make this a character "role" like people can be medical doctor and traitor, a neutral player modifier of some sort

**Comment by Jarmer123** [comment]
As one of the admins this sounds like a real pain to moderate and I don't really get what this role is suppose to do except for give someone dead another chance to just be a crew member that now hears OOC chat, like most mid round spawns have an objective. I think it could be interesting if it was expanded a bit.

**Comment by c4llv07e** [comment]
> Hobo is somewhat derogatory, another name would be preferable, e.g., vagrant or drifter. Or you could lean into the ghost aspects and call them a medium, seer, psychic, etc

My fault, I didn't even think about it. (changed)

File: Content.Server/Chat/Systems/ChatSystem.cs [hunk 1]
```diff
@@ -588,7 +588,14 @@ private void SendLOOC(EntityUid source, ICommonSession player, string message, b
...
     private void SendDeadChat(EntityUid source, ICommonSession player, string message, bool hideChat)
     {
-        var clients = GetDeadChatClients();
+        var dead_clients = GetDeadChatClients();
+        var clients = dead_clients.Concat(
+            GetRecipients(source, VoiceRange)
+            .Select(p => p.Key)
+            .Where(p => HasComp<DeadHearingComponent>(p.AttachedEntity))
+            .Select(p => p.Channel)
+            .Except(dead_clients)); // Remove duplicates
```
Comment (metalgearsloth) [inline Content.Server/Chat/Systems/ChatSystem.cs:597]: Please don't abuse LINQ like this it's incredibly slow.

**Comment by IProduceWidgets** [comment]
I agree with others, the role needs a catch / objective / drawback. As is its just a respawn with OOC info and metagaming. Sounds like a massive pain to admin, and I see no benefit to the round.

I like the suggestion of a word whitelist, but I'm not sure how you'll communicate it to the ghosts. Still though, it needs a bigger drawback, like the player having the gibberish accent component(s) or something.

File: Content.Server/Chat/Systems/ChatSystem.cs (hunk 1 above)
Comment (deltanedas) [reply Content.Server/Chat/Systems/ChatSystem.cs:597]: linq

**Comment by deltanedas** [comment]
> like the player having the gibberish accent component(s) or something.

like nrp mime pros will just write "nukies parked outside armoury" on paper then spam point at captain

**Comment by IProduceWidgets** [comment]
true, but at least its not just yelling over radio :shrugs:

**Comment by mirrorcult** [comment]
make their speech gibberish and unable to write

**Comment by deltanedas** [comment]
now drifter will spam point at comms console and get command access, then announce their meta info

"you can have all this meta info from deadchat but pls dont use it in any way shape or form" is a bit silly

**Comment by mirrorcult** [comment]
this isnt "pls dont use it" its "make it impossible to communicate it to anyone". make them unable to use consoles then

**Comment by deltanedas** [comment]
better make them unable to drops items too since they could write on the floor with them
better make them unable to use crayons since they literally have letters
better make them unable to build too since they could write with enough materials (bit of a stretch)

its easiest to just not give them info, if they cant use it then why give it in the first place

**Comment by mirrorcult** [comment]
they should be able to use it, they just shouldnt be able to communicate to anyone else about it

**Comment by mirrorcult** [comment]
also yes, do prohibit them from drawing with crayons, the other things dont matter that much though

**Comment by Aisu9** [comment]
they will be able to use the communication console in bridge if the captain or anyone in command ever let them in and/or ever get "AA" from HoP

**Comment by Ilya246** [comment]
they will be able to draw letters with pointing

**Comment by mirrorcult** [comment]
good luck with that

**Comment by deltanedas** [comment]
if you see a space drifter making bolas and shit you could infer that its nukies

"hey drifter nice bolas, are nukies coming?"
@nods
":c drifter says nukies"

all without directly leaking deadchat

**Comment by Aisu9** [comment]
even if the space drifter couldn't hear ghost and speak the ghost(s) would just "draw" words with their movement or their number

**Comment by mirrorcult** [comment]
bro ghosts would not be coordinated enough to do that are you insane

**Comment by Aisu9** [comment]
we all know the players are very talented when it come to breaking or bypassing limitation, i don't give them 24h before they managed to find something

**Comment by icekot8** [comment]
Pr makes sense, but I don't understand a few things.

1. What's the point of it? Like, a dude just shows up and just talks to ghosts? Seriously? Why... It seems like a meta accessibility of someone who can't spread about it, I don't mind being able to hear ghosts, but listening to GLOBAL chat of the dead-subjectively isn't right, too much meta and non rp information. I think it would be fair to make goals for him where ghosts will help him in some way, maybe one ghost that gets a "mentor" role... It's not clear to me why he can see and hear ghosts if he can't use information from them. meta fear exists.
2. Will he see the revenant? If so, a non-antagonist can see an antagonist capable of killing the medical department morally. If no-it's not clear how the revenant differs from ghosts, essentially-he's a spirit too
3. Will he have a ghost disable button? At important moments can gather crowds of ghosts, which will banal interfere with space drifters

I think the pr needs some serious changes.....

**Comment by deltanedas** [comment]
> revenant
> an antagonist capable of killing

:trollface:

**Comment by Ilya246** [comment]
the drifter should probably be an antag with some objective so that it doesn't cooperate with the crew, but then it would also need a design doc

also, add your components onto revenant, revenant is supposed to be able to hear ghosts

**Comment by deltanedas** [comment]
revenant hearing ghosts talk about nukies might be useful since it can be strategic and camp nukie kills

**Comment by Ilya246** [comment]
> revenant hearing ghosts talk about nukies might be useful since it can be strategic and camp nukie kills

yes good

**Comment by IProduceWidgets** [comment]
> now drifter will spam point at comms console and get command access, then announce their meta info
> 
> "you can have all this meta info from deadchat but pls dont use it in any way shape or form" is a bit silly

I don't think it needs to be completely impossible to communicate it, just like, difficult enough to be both possibly misinterpreted and also not worth it to try to pass on every little thing a ghost does.

after having thought about it for a bit though, it might be better if the drifter can only see ghosts, and not hear them coherently. Spooky gasping or 'boooooOooooOoooo's though seem fine.

**Comment by metalgearsloth** [comment]
Ping if you come back to this.

**Pull request closed without merging**